File: facebook-core/src/main/java/com/facebook/internal/qualityvalidation/ExcusesForDesignViolations.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.internal.qualityvalidation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Annotations to mark code files as having known design issues and avoid failing the design
 * integration test on them
 *
 * <p>Please avoid using it for everything that can be fixed trivially, and delaying fixes to future
 * diffs without a good reason.
 *
 * <p>To use it list one of the type of issue you want to ignore, and supply and extra string
 * explaining why this is okay for this case.
 *
 * <pre>{@code
 * @ExcusesForDesignViolations(
 *   @Excuse(type = "STORY_VIEWER_CAPITALIZATION", reason = "Legacy"),
 *   @Excuse(type = "MISSING_JAVA_DOC", reason = "DI Module")
 * )
 * }</pre>
 */
@Retention(RetentionPolicy.SOURCE)
@Target(ElementType.TYPE)
public @interface ExcusesForDesignViolations {
  Excuse[] value();
}

File: facebook-core/src/main/java/com/facebook/internal/qualityvalidation/ExcusesForDesignViolations.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal.qualityvalidation

import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy

/**
 * Annotations to mark code files as having known design issues and avoid failing the design
 * integration test on them
 *
 * <p>Please avoid using it for everything that can be fixed trivially, and delaying fixes to future
 * diffs without a good reason.
 *
 * <p>To use it list one of the type of issue you want to ignore, and supply and extra string
 * explaining why this is okay for this case.
 *
 * <pre>{@code
 * @ExcusesForDesignViolations(
 * @Excuse(type = "STORY_VIEWER_CAPITALIZATION", reason = "Legacy"),
 * @Excuse(type = "MISSING_JAVA_DOC", reason = "DI Module") ) }</pre>
 */
@Retention(RetentionPolicy.SOURCE)
@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
annotation class ExcusesForDesignViolations(vararg val value: Excuse)
